packages/bundlers/default/src/DefaultBundler.js

```diff
@@ -160,14 +160,54 @@ export default new Bundler({
   },
 
   optimize({bundleGraph}) {
-    // Step 2: remove assets that are duplicated in a parent bundle
+    // Step 2: Remove asset graphs that begin with entries to other bundles.
+    bundleGraph.traverseBundles(bundle => {
+      if (bundle.isInline || !bundle.isSplittable) {
+        return;
+      }
+
+      let mainEntry = bundle.getMainEntry();
+      if (mainEntry == null) {
+        return;
+      }
+
+      let siblings = bundleGraph
+        .getSiblingBundles(bundle)
+        .filter(sibling => !sibling.isInline);
+      let candidates = bundleGraph.findBundlesWithAsset(mainEntry).filter(
+        containingBundle =>
+          containingBundle.id !== bundle.id &&
+          // Don't add to BundleGroups for entry bundles, as that would require
+          // another entry bundle depending on these conditions, making it difficult
+          // to predict and reference.
+          !containingBundle.isEntry &&
+          !containingBundle.isInline &&
+          containingBundle.isSplittable,
+      );
+
+      for (let candidate of candidates) {
+        let bundleGroups = bundleGraph.getBundleGroupsContainingBundle(
+          candidate,
+        );
+        if (!exceedsParallelRequestLimit(bundleGraph, bundleGroups)) {
+          bundleGraph.removeAssetGraphFromBundle(mainEntry, candidate);
+          for (let bundleGroup of bundleGroups) {
+            for (let bundleToAdd of [bundle, ...siblings]) {
+              bundleGraph.addBundleToBundleGroup(bundleToAdd, bundleGroup);
```

Comment:
My only concern using async/entry bundles as shared bundles is that these bundles have asset entries — code that will run immediately once they are evaluated — unlike the shared bundles created below. This doesn't seem to have been a problem so far. What do you think, @devongovett?